### PR TITLE
[config] make configuration value parser available for export

### LINKED ISSUE #2700: [config] Expose ConfigMapper to streamline parsing of configuration parameters
[...]

Looking into the `ConfigMapper` class I am wondering why we did not expose it and the static method `objectConvert(Object, Class<?>)` with it to reuse it everywhere where we have to parse configuration parameters e.g. to prevent issues like https://github.com/openhab/openhab-core/issues/2687. The mapping util can be used for OSGi service configurations too.

_Originally posted by @cweitkamp in https://github.com/openhab/openhab-core/issues/2699#issuecomment-1017776072_

### PR DESCRIPTION
**Pull request by J-N-K**
Closes #2700 

Some points are still open:

- If a conversion is not possible (e.g. provided value is `"foo"` and desired type is `int`), shall we return `null` or throw an `IllegalArgumentException` (or better a newly defined checked `ConfigValueParserException`)?
- Currently `float`, `int`, `long` and `double` (and their wrapper classes) are supported. Shall we add code for `short`, `byte` and `char`?
- There is no support for arrays. Is that needed?
- All `Collections` are mapped to `List`. Shall we at least add `Set`?
-

### CODE REVIEW

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java [hunk 1]
```diff
@@ -120,49 +141,81 @@
...
-    private static Object objectConvert(Object value, Class<?> type) {
+    /**
+     * Convert a value to a given type
+     *
+     * @param value input value or String representation of that value
+     * @param type desired target class
+     * @param defaultValue default value to be used if conversion fails or input value is null
+     * @return the converted value
+     */
+    public static <T> T valueAs(@Nullable Object value, Class<T> type, T defaultValue) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java:152]: We can name it `valueAsElse()` or `valueAsOrElse()` to let the code have a better readability.

```suggestion
    public static <T> T valueAsElse(@Nullable Object value, Class<T> type, T defaultValue) {
```

**Review by cweitkamp — COMMENTED** [review]
I think we should not throw exceptions rather log a warning. I assume most of the consumers have a proper `null` handling - if they do not use the `valueAsElse()` method - thus exception handling will lead to boilerplate code to be added again. The idea behind this PR is to reduce code.

Adding support for `short`, `byte` or `Set` does not hurt. Imo arrays are not needed.

**Comment by J-N-K** [comment]
Ready for review from my side. I would suggest to adjust code to use this in a follow-up PR instead of changing it here.

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    /**
+     * Convert a value to a given type
+     *
+     * @param value input value or String representation of that value
+     * @param type desired target class
+     * @return the converted value or null if conversion fails or input value is null
+     */
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public static <T> @Nullable T valueAs(@Nullable Object value, Class<T> type) {
+        // make sure primitives are converted to their respective wrapper class
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java:178]: Should we add an early return if the` value` already is an instance of the requested `type`?

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    /**
+     * Use this method to automatically map a configuration collection to a Configuration holder object. A common
+     * use-case would be within a service. Usage example:
+     *
+     * <pre>
+     * {@code
+     *   public void modified(Map<String, Object> properties) {
+     *     YourConfig config = ConfigMapper.as(properties, YourConfig.class);
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java:65]:
```suggestion
     *     YourConfig config = ConfigParser.configurationAs(properties, YourConfig.class);
```

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            Class<?> type = field.getType();
+
+            Object value = properties.get(fieldName);
+            // Consider RequiredField annotations
+            if (value == null) {
+                LOGGER.trace("Skipping field '{}', because config has no entry for {}", fieldName, fieldName);
+                continue;
+            }
+
+            // Allows to have List<int>, List<Double>, List<String> etc (and the corresponding Set<?>
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java:104]:
```suggestion
            // Allows to have List<int>, List<Double>, List<String> etc (and the corresponding Set<?>)
```

File: bundles/org.openhab.core.config.core/src/test/java/org/openhab/core/config/core/ConfigParserTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            new TestParameter<>(1, Byte.class, (byte) 1), //
+            // boolean/Boolean
+            new TestParameter<>("true", boolean.class, true), //
+            new TestParameter<>("true", Boolean.class, true), //
+            new TestParameter<>(false, boolean.class, false), //
+            new TestParameter<>(false, Boolean.class, false), //
+            // BigDecimal
+            new TestParameter<>("7.5", BigDecimal.class, BigDecimal.valueOf(7.5)), //
+            new TestParameter<>(BigDecimal.valueOf(-7.5), BigDecimal.class, BigDecimal.valueOf(-7.5)), //
+            new TestParameter<>(1, BigDecimal.class, BigDecimal.valueOf(1)), //
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/test/java/org/openhab/core/config/core/ConfigParserTest.java:74]:
```suggestion
            new TestParameter<>(1, BigDecimal.class, BigDecimal.ONE), //
```

File: bundles/org.openhab.core.config.core/src/test/java/org/openhab/core/config/core/ConfigurationTest.java [hunk 6]
```diff
@@ -62,6 +66,7 @@ public void assertGetConfigAsWorks() {
         configuration.put("stringField", "test");
         configuration.put("enumField", "ON");
         configuration.put("listField", List.of("one", "two", "three"));
+        configuration.put("setField", List.of("one", "two", "three"));
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/test/java/org/openhab/core/config/core/ConfigurationTest.java:69]: Shouldn't thi be as `Set`?

```suggestion
        configuration.put("setField", Set.of("one", "two", "three"));
```

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        // make sure primitives are converted to their respective wrapper class
+        Class<?> typeClass = WRAPPER_CLASSES_MAP.getOrDefault(type.getSimpleName(), type);
+
+        Object result = value;
+        // Handle the conversion case of BigDecimal to Float,Double,Long,Integer
+        if (value instanceof BigDecimal && !BigDecimal.class.equals(typeClass)) {
+            BigDecimal bdValue = (BigDecimal) value;
+            if (Float.class.equals(typeClass)) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java:190]: This `if`-`else`-statement looks very similar to the block below and `BigDecimal` implements `Number` too thus I am wondering if we can reduce code by extracting a private method for it. Do you see any chance to apply that?

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        return fields;
+    }
+
+    /**
+     * Convert a value to a given type or return default value
+     *
+     * @param value input value or String representation of that value
+     * @param type desired target class
+     * @param defaultValue default value to be used if conversion fails or input value is null
+     * @return the converted value
```
Comment (cweitkamp) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java:163]:
```suggestion
     * @return the converted value or the given defaultValue if conversion fails or input value is null
```

**Review by cweitkamp — COMMENTED** [review]
Thanks for your work. The feature looks really nice. Cannot wait to use it in different places.

I read through the code on last time and left some final remarks.

File: bundles/org.openhab.core.config.core/src/test/java/org/openhab/core/config/core/ConfigurationTest.java (hunk 6 above)
Comment (J-N-K) [reply bundles/org.openhab.core.config.core/src/test/java/org/openhab/core/config/core/ConfigurationTest.java:69]: No, I deliberately chose to use List as input to check if a List is properly converted to a Set if that is required. Maybe another Set->Set and Set->List test should be added.

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java (hunk 7 above)
Comment (J-N-K) [reply bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/ConfigParser.java:190]: The first `BigDecimal` case can be removed in total, it's covered by the previous check if the type is already ok and the following `Number`.

**Comment by J-N-K** [comment]
Tests fails unrelated

**Review by cweitkamp — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**